From 30cfd41d677bc746fca701e35162b93953430b27 Mon Sep 17 00:00:00 2001
From: "Valery V. Vorotyntsev" <valery_vorotyntsev@xyratex.com>
Date: Sat, 17 Mar 2012 00:44:55 +0200
Subject: [PATCH 069/311] c2_fop_type_format_parse(): document; fix a bug

Add doxygen comment describing what c2_fop_type_format_parse() does.
Handle NULL returned by C2_ALLOC_ARR().
---
 fop/fop_format.c | 6 ++++--
 fop/fop_format.h | 6 ++++++
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/fop/fop_format.c b/fop/fop_format.c
index f11d33d..0286310 100644
--- a/fop/fop_format.c
+++ b/fop/fop_format.c
@@ -145,7 +145,9 @@ int c2_fop_type_format_parse(struct c2_fop_type_format *fmt)
 	result = 0;
 	if (nr > 0) {
 		C2_ALLOC_ARR(t->fft_child, nr);
-		if (t->fft_child != NULL) {
+		if (t->fft_child == NULL) {
+			result = -ENOMEM;
+		} else {
 			for (i = 0; i < nr; ++i) {
 				struct c2_fop_field              *field;
 				const struct c2_fop_field_format *field_fmt;
@@ -177,8 +179,8 @@ int c2_fop_type_format_parse(struct c2_fop_type_format *fmt)
 	}
 
 	if (result == 0) {
-		fmt->ftf_out = t;
 		t->fft_layout = fmt->ftf_layout;
+		fmt->ftf_out = t;
 	} else
 		c2_fop_type_format_fini(fmt);
 	return result;
diff --git a/fop/fop_format.h b/fop/fop_format.h
index 539b997..333f114 100644
--- a/fop/fop_format.h
+++ b/fop/fop_format.h
@@ -96,7 +96,13 @@ struct c2_fop_type_format {
 	} ftf_child[];
 };
 
+/**
+ * Allocates and initialises fmt->ftf_out (c2_fop_field_type).
+ *
+ * @pre fmt->ftf_out == NULL
+ */
 int  c2_fop_type_format_parse(struct c2_fop_type_format *fmt);
+
 void c2_fop_type_format_fini(struct c2_fop_type_format *fmt);
 
 int  c2_fop_type_format_parse_nr(struct c2_fop_type_format **fmt, int nr);
-- 
1.8.3.2

